provider/lib/app.dart

```diff
@@ -0,0 +1,49 @@
+// Copyright 2018 The Flutter Architecture Sample Authors. All rights reserved.
+// Use of this source code is governed by the MIT license that can be found
+// in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:meta/meta.dart';
+import 'package:provider/provider.dart';
+import 'package:provider_sample/add_todo_screen.dart';
+import 'package:provider_sample/localization.dart';
+import 'package:provider_sample/todo_list_model.dart';
+import 'package:todos_app_core/todos_app_core.dart';
+import 'package:todos_repository_core/todos_repository_core.dart';
+
+import 'home/home_screen.dart';
+
+class ProviderApp extends StatelessWidget {
+  final TodosRepository repository;
+
+  ProviderApp({
+    @required this.repository,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return ChangeNotifierProvider(
+      create: (_) => TodoListModel(repository: repository),
+      child: MaterialApp(
+        title: ProviderLocalizations().appTitle,
+        theme: ArchSampleTheme.theme,
+        localizationsDelegates: [
+          ArchSampleLocalizationsDelegate(),
+          ProviderLocalizationsDelegate(),
+        ],
+        routes: {
+          ArchSampleRoutes.home: (context) => HomeScreen(),
+          ArchSampleRoutes.addTodo: (context) {
+            return AddTodoScreen(
+              onAdd: (todo) {
```

Comment:
This probably should be a method on `TodoListModel` instead, to concentrate the logic in one place (unless you don't want it to use `provider`?)

For the `Navigator.pop`, we have a few alternatives:
- moved to `AddTodoScreen` next to the current `onAdd` call (the most logical IMO)
- make a NavigatorService and provide it , then pass it to `TodoListModel` – a bit overkill and I don't quite like it
- have an `onAdd` callback on `TodoListModel` instead

Reply:
Hum seeing the `showSnackbar` in home, which can't be refactored the same way, I'm now uncertain.

An alternative that would work with `Home`/`showSnackbar` is to use `Notification`/`NotificationListener`.

It's making things a more complex though. So keeping it this way is not too unreasonable. Thought?

Reply:
> This probably should be a method on TodoListModel instead, to concentrate the logic in one place (unless you don't want it to use provider?)

Yah, I don't think we can concentrate this logic in one place, other than the `addTodo` portion. The Add Screen needs to add a todo then pop the navigator. The Snackbar needs to just add the todo. The edit screen needs to update then pop, whereas checkboxes just need to update the todo but not pop.

Therefore, I think the current structure makes the most sense. Provide the generic `add` and `update` methods on the Model, which is responsible for managing the todos, and let the Widgets decide how to use those methods and control the visuals themselves.